helion/autotuner/cache.py

```diff
@@ -0,0 +1,121 @@
+
+log: logging.Logger = logging.getLogger(__name__)
+
+"""
+TODO(oulgen)
+- Allow user defined cache keys that can be passed on @helion.kernel
+- Add import/export for set of configs
+"""
+
+
+@dataclasses.dataclass(frozen=True)
+class AutotuneCacheKey:
+    """
+    helion_key: Hash of source code of Helion
+    torch_key: Hash of source code of PyTorch
+    system_hash: Hash of system information,
+        including Triton, current device, cuda/rocm arch version
+    function_source_hash: Hash of source code of input Helion kernel
+    input_dtypes: dtypes of input tensors
+    input_shapes: shapes of input tensors
+    """
+
+    helion_key: str
+    torch_key: str
+    system_hash: str
+    kernel_source_hash: str
+    input_dtypes: list[tuple[int, torch.dtype]]
+    input_shapes: list[tuple[int, torch.Size]]
+
+    def stable_hash(self) -> str:
+        return hashlib.sha256(repr(self).encode("utf-8")).hexdigest()
+
+
+class AutotuneCache:
+    def __init__(self, kernel: Kernel, args: Sequence[object]) -> None:
+        self.key: AutotuneCacheKey = AutotuneCache._generate_key(kernel, args)
+
+    @staticmethod
+    def _generate_key(kernel: Kernel, args: Sequence[object]) -> AutotuneCacheKey:
+        from torch._inductor.codecache import CacheBase
+        from torch._inductor.codecache import torch_key
+
+        kernel_source = textwrap.dedent(inspect.getsource(kernel.fn))
+        kernel_source_hash = hashlib.sha256(kernel_source.encode("utf-8")).hexdigest()
+
+        input_dtypes = []
+        input_shapes = []
+
+        for idx, a in enumerate(args):
+            if isinstance(a, torch.Tensor):
+                input_dtypes.append((idx, a.dtype))
+                input_shapes.append((idx, a.shape))
+
+        return AutotuneCacheKey(
+            helion_key=helion_key(),
+            torch_key=torch_key().hex(),
+            system_hash=CacheBase.get_system()["hash"],
+            kernel_source_hash=kernel_source_hash,
+            input_dtypes=input_dtypes,
+            input_shapes=input_shapes,
+        )
+
+    def _get_cache_key(self) -> str:
+        return self.key.stable_hash()
+
+    def _get_local_cache_path(self) -> Path:
+        from torch._inductor.runtime.cache_dir_utils import (
+            cache_dir,  # pyright: ignore[reportPrivateImportUsage]
+        )
+
+        return Path(cache_dir()) / "helion" / f"{self._get_cache_key()}.best_config"
+
+    def get(self) -> Config | None:
+        path = self._get_local_cache_path()
+        try:
+            config = Config.load(path)
+            log.debug("Cache hit on config at %s", path)
+            counters["autotune"]["cache_hit"] += 1
+            return config
+        except Exception:
+            log.debug("Cache miss on config at %s", path)
+            counters["autotune"]["cache_miss"] += 1
+            return None
+
+    def put(self, config: Config) -> None:
+        path = self._get_local_cache_path()
+        config.save(path)
```

Comment:
Should we do the save to a temp file and rename trick here so the change is atomic.  I thinking about multi-process races.